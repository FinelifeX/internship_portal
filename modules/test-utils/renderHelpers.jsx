/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import TR from 'react-test-renderer';
import { render as RTLRender } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

/**
 * @typedef {Object.<string, any>} Props
 */

/**
 * @typedef {React.ReactNode} Children
 */

/**
 * @typedef {{props?: Props, children?: Children}} RenderOptions
 */

/**
 *
 * @param {React.ComponentType} Component
 * @param {RenderOptions} [options]
 * @returns {JSX.Element}
 */
export const render = (Component, options = {}) => {
  const { props, children } = options;
  return <Component {...props}>{children}</Component>;
};

/**
 * Render React component inside MemoryRouter
 * @param {React.ComponentType} Component
 * @param {RenderOptions} [options]
 * @param {import('react-router-dom').MemoryRouterProps} routerOptions
 * @returns {JSX.Element}
 */
export const renderWithRouter = (
  Component,
  options = {},
  routerOptions = {},
) => {
  const { props, children } = options;

  return (
    <MemoryRouter {...routerOptions}>
      <Component {...props}>{children}</Component>
    </MemoryRouter>
  );
};

/**
 * Shallow render component using "react-test-renderer"
 * @type {import('react-test-renderer').create}
 */
export const shallow = TR.create;

/**
 * Render component using "render()" from "@testing-library/react"
 * @type {import('@testing-library/react').render}
 */
export const mount = RTLRender;
