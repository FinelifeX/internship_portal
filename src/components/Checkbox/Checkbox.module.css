.container {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  cursor: pointer;
}

.checkbox {
  margin: 0 8px 0 0;
  height: 16px;
  width: 16px;
  cursor: pointer;
}

.label {
  font-size: 14px;
}
