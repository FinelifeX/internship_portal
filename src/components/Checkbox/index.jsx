import React from 'react';
import PT from 'prop-types';
import classNames from 'classnames';
import styles from './Checkbox.module.css';

const Checkbox = ({ label, className, checked, onChange }) => (
  <span className={classNames(styles.container, className)}>
    <input
      data-testid="checkbox"
      className={styles.checkbox}
      id="checkbox"
      type="checkbox"
      checked={checked ?? undefined}
      onChange={onChange}
    />
    <label data-testid="label" className={styles.label} htmlFor="checkbox">
      {label}
    </label>
  </span>
);

Checkbox.defaultProps = {
  className: '',
  checked: undefined,
  onChange: undefined,
};

Checkbox.propTypes = {
  label: PT.string.isRequired,
  className: PT.string,
  checked: PT.bool,
  onChange: PT.func,
};

export default Checkbox;
