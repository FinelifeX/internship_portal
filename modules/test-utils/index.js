import * as RenderHelpers from './renderHelpers';

export { RenderHelpers };
