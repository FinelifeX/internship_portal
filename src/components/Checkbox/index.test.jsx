import { fireEvent } from '@testing-library/react';
import { domRender, shallowRender } from 'test-utils/renderHelpers';
import Checkbox from './index';

describe('Checkbox component', () => {
  const TEST_ID_INPUT = 'checkbox';
  const TEST_ID_LABEL = 'label';
  const TEST_ID_TEXT = 'text';

  const defaultProps = {
    label: 'checkbox',
    checked: true,
    onChange: jest.fn(),
  };

  it('should render correctly', () => {
    const elem = shallowRender(Checkbox, {
      defaultProps,
    }).toJSON();

    expect(elem).toMatchSnapshot();
  });

  const renderComponent = (props) =>
    domRender(Checkbox, { defaultProps, props });

  it('should have correct label', () => {
    const { getByTestId } = renderComponent();

    expect(getByTestId(TEST_ID_TEXT)).toHaveTextContent(defaultProps.label);
  });

  it('should be checked if checked prop === true', () => {
    const props = {
      checked: true,
    };
    const { getByTestId } = renderComponent(props);

    expect(getByTestId(TEST_ID_INPUT)).toBeChecked();
  });

  it('should call onChange 1 time when checkbox is clicked', () => {
    const { getByTestId } = renderComponent();
    const input = getByTestId(TEST_ID_INPUT);

    fireEvent.click(input);

    expect(defaultProps.onChange).toBeCalledTimes(1);
  });

  it("should call onChange when checkbox's label was clicked", () => {
    const { getByTestId } = renderComponent();
    const label = getByTestId(TEST_ID_LABEL);

    fireEvent.click(label);

    expect(defaultProps.onChange).toBeCalled();
  });

  it('should be checked after being clicked', () => {
    const { getByTestId } = renderComponent();
    const elem = getByTestId(TEST_ID_INPUT);

    fireEvent.click(elem);

    expect(elem).toBeChecked();
  });
});
