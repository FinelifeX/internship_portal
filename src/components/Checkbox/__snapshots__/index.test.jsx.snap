// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component should render correctly 1`] = `
<label
  className="container"
  data-testid="label"
  htmlFor="checkbox"
>
  <input
    checked={true}
    className="checkbox"
    data-testid="checkbox"
    id="checkbox"
    onChange={[MockFunction]}
    type="checkbox"
  />
  <span
    className="label"
    data-testid="text"
  >
    checkbox
  </span>
</label>
`;
